connaisseur/validate.py

```diff
@@ -92,11 +94,17 @@ def process_chain_of_trust(host: str, image: Image, req_delegations: list):
     # as well
     if trust_data["targets"].has_delegations():
         for delegation in trust_data["targets"].get_delegations():
-            trust_data[delegation] = get_trust_data(host, image, TUFRole(delegation))
+            trust_data[delegation] = get_delegation_trust_data(
+                host, image, TUFRole(delegation)
+            )
 
-    # validate all trust data's signatures, expiry dates and hashes
+    # validate all trust data's signatures, expiry dates and hashes.
+    # when delegation are added to the repository, but weren't yet used for signing, the
+    # delegation files don't exist yet and are `None`. in this case they can't be
+    # validated and must be skipped.
     for role in trust_data:
-        trust_data[role].validate(key_store)
+        if trust_data[role] is not None:
```

Comment:
What happens if an attacker controlling the notary server deletes e.g. the `timestamp.json`? Is the verification of that file skipped in that case?

Reply:
All the mayor trust data (root, targets, timestamp, snapshot) will be retrieved using the `get_trust_data()` function, which will raise an error if the corresponding trust data can't be found. Meaning they can't be `None` there.

Even if the attacker just deletes the content of the file, while still keeping it, a `TrustData` object is still created, just with empty data. So even in this case it shouldn't be `None`. Only the delegation trust data can be `None`, but those are further checked, should specific delegations be needed, or ignored if not needed.

Reply:
thanks!

Reply:
If you find yourself bored, you could create testcases for the behavior that occurs when trust data files are partially missing etc.

Reply:
The point being that any later refactoring/change would be prevented from introducing the kind of bug @annekebr is worried about.